central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -231,16 +429,72 @@ func (ds *datastoreImpl) DryRunAddCollection(ctx context.Context, collection *st
 	return ds.dryRunAddCollectionInGraph(collection)
 }
 
+func (ds *datastoreImpl) UpdateCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+	if err := verifyCollectionObjectNotEmpty(collection); err != nil {
+		return err
+	}
+
+	// resolve object to check if the name was changed
+	obj, ok, err := ds.storage.Get(ctx, collection.GetId())
+	if err != nil || !ok {
+		return errors.Wrap(err, "failed to resolve collection being updated")
+	}
+	if obj.GetName() != collection.GetName() && ds.names.Contains(collection.GetName()) {
+		return fmt.Errorf("collection name already in use (%s)", collection.GetName())
+	}
+
```

Comment:
Hold the lock here. Holding the lock inside `updateCollectionInGraphWorkflow` means once the graph is updated, the lock is released and the next op—add, update, delete—can continue even before we have made state changes to db. Plus, it will lead to error when you undo below.

With that, you can change the signature of `updateCollectionInGraphWorkflow` to `dryRunUpsertCollectionInGraph(obj *storage.ResourceCollection) (*dag.DAG, error)`. You set `ds.graph = graphCopy` if the db operation succeeds. I believe with this change, `undoEdgeUpdatesInGraph` is no longer needed.